File: go/mariani-silver/imageGenerator.go
package main

import "sync"

const maxRectangleSize = 50

type ImageGenerator struct {
	calculator MandelbrotPixelCalculator
}

func NewImageGenerator() ImageGenerator {
	return ImageGenerator{
		NewMandelbrotPixelCalculator(),
	}
}

type ImageGenerationContext struct {
	calculator MandelbrotPixelCalculator
	image      *Image
	dimensions ImageDimensions
	waitGroup  sync.WaitGroup
}

func (generator ImageGenerator) generate(image *Image, dimensions ImageDimensions) {
	assertImageMatchesDimensions(image, dimensions)
	context := ImageGenerationContext{
		generator.calculator,
		image,
		dimensions,
		sync.WaitGroup{},
	}
	fullRectangle := rectangleFromDimensions(dimensions)
	context.workOn(fullRectangle)
	context.waitGroup.Wait()
}

func (context *ImageGenerationContext) workOn(rectangle Rectangle) {
	if context.isTiny(rectangle) {
		context.generateWholeRectangle(rectangle)
	} else {
		context.calculateOutlinesAndSplit(rectangle)
	}
}

func (context *ImageGenerationContext) workOnDone(rectangle Rectangle) {
	context.workOn(rectangle)
	context.waitGroup.Done()
}

func (context *ImageGenerationContext) generateWholeRectangle(rectangle Rectangle) {
	for x := rectangle.x; x < rectangle.endX; x++ {
		for y := rectangle.y; y < rectangle.endY; y++ {
			context.generatePoint(x, y)
		}
	}
}

func (context *ImageGenerationContext) calculateOutlinesAndSplit(rectangle Rectangle) {
	outlinesIdentical, referenceColor := context.generateOutlines(rectangle)

	if outlinesIdentical {
		context.fillWholeRectangle(rectangle, referenceColor)
	} else {
		context.split(rectangle)
	}
}

func (context *ImageGenerationContext) generateOutlines(
	rectangle Rectangle,
) (outlinesIdentical bool, referenceColor Pixel) {
	topY := rectangle.y
	bottomY := rectangle.endY - 1
	leftX := rectangle.x
	rightX := rectangle.endX - 1

	outlinesIdentical = true
	referenceColor = context.generatePoint(leftX, topY)

	for x := leftX; x <= rightX; x++ {
		context.checkOutlinePixel(x, topY, referenceColor, &outlinesIdentical)
		context.checkOutlinePixel(x, bottomY, referenceColor, &outlinesIdentical)
	}
	for y := topY; y <= bottomY; y++ {
		context.checkOutlinePixel(leftX, y, referenceColor, &outlinesIdentical)
		context.checkOutlinePixel(rightX, y, referenceColor, &outlinesIdentical)
	}
	return outlinesIdentical, referenceColor
}

func (context *ImageGenerationContext) checkOutlinePixel(x, y uint, referenceColor Pixel, outlinesIdentical *bool) {
	currentColor := context.generatePoint(x, y)
	if currentColor != referenceColor {
		*outlinesIdentical = false
	}
}

func (context *ImageGenerationContext) split(rectangle Rectangle) {
	parts := SplitRectangle(rectangle)
	for _, part := range parts {
		context.waitGroup.Add(1)
		go context.workOnDone(part)
	}
}

func (context *ImageGenerationContext) fillWholeRectangle(rectangle Rectangle, color Pixel) {
	for x := rectangle.x; x < rectangle.endX; x++ {
		for y := rectangle.y; y < rectangle.endY; y++ {
			target := context.image.Get(x, y)
			*target = color
		}
	}
}

func (context *ImageGenerationContext) generatePoint(x, y uint) Pixel {
	empty := Pixel{}
	target := context.image.Get(x, y)
	if *target != empty {
		return *target
	}
	sampled := context.dimensions.SampleTo(x, y)
	color := context.calculator.GetColorFor(sampled)
	*target = color
	return color
}

func (context *ImageGenerationContext) isTiny(rectangle Rectangle) bool {
	return rectangle.width <= maxRectangleSize && rectangle.height <= maxRectangleSize
}

func assertImageMatchesDimensions(image *Image, dimensions ImageDimensions) {
	if image.height != dimensions.height || image.width != dimensions.width {
		panic("Dimensions do not match")
	}
}

func rectangleFromDimensions(dimensions ImageDimensions) Rectangle {
	return NewRectangle(
		0, 0,
		dimensions.width, dimensions.height,
	)
}

File: rust/mariani-silver/src/image_generator.rs
pub mod image_generator {
	use std::sync::Arc;
	use std::sync::atomic::{AtomicUsize, Ordering};
	use std::time::Duration;

	use async_std::task;

	use crate::image::image::{Image, Pixel};
	use crate::image_dimensions::image_dimensions::ImageDimensions;
	use crate::mandelbrot_pixel_calculation::mandelbrot_pixel_calculation::MandelbrotPixelCalculator;
	use crate::rectangle::rectangle::Rectangle;
	use crate::rectangle_splitter::rectangle_splitter::split_rectangle;
	use crate::wait_group::wait_group::WaitGroup;

	const MAX_RECTANGLE_SIZE: usize = 50;

	#[derive(Clone)]
	pub struct GenerationContext {
		calculator: Arc<MandelbrotPixelCalculator>,
		image: Arc<Image>,
		dimensions: Arc<ImageDimensions>,
	}

	pub fn generate(
		calculator: Arc<MandelbrotPixelCalculator>,
		image: Arc<Image>,
		dimensions: Arc<ImageDimensions>,
	) {
		assert_image_matches_dimensions(image.as_ref(), dimensions.as_ref());
		let full_rectangle = rectangle_from_dimensions(dimensions.as_ref());
		let context = GenerationContext {
			calculator,
			image,
			dimensions,
		};
		let wait_group = WaitGroup::new();
		generate_within(
			context.clone(), full_rectangle, wait_group.clone(),
		);
		wait_group.wait()
	}

	fn generate_within(
		context: GenerationContext,
		rectangle: Rectangle,
		wait_group: WaitGroup,
	) {
		if is_tiny(&rectangle) {
			generate_whole_rectangle(context, &rectangle);
		} else {
			calculate_outlines_and_split(context, &rectangle, wait_group);
		}
	}

	fn generate_whole_rectangle(
		context: GenerationContext,
		rectangle: &Rectangle,
	) {
		for x in rectangle.x..rectangle.end_x {
			for y in rectangle.y..rectangle.end_y {
				generate_point(
					context.clone(),
					x,
					y,
				);
			}
		}
	}

	fn generate_point(
		context: GenerationContext,
		x: usize,
		y: usize,
	) -> Pixel {
		const EMPTY: Pixel = [0u8, 0u8, 0u8];
		let target = context.image.index_mut_unsafe(x, y);
		if *target != EMPTY {
			return *target;
		}
		let sampled = context.dimensions.sample_to(x, y);
		let color = context.calculator.get_color_for(sampled);
		*target = color;
		*target
	}

	fn is_tiny(rectangle: &Rectangle) -> bool {
		rectangle.width <= MAX_RECTANGLE_SIZE && rectangle.height <= MAX_RECTANGLE_SIZE
	}

	fn calculate_outlines_and_split(
		context: GenerationContext,
		rectangle: &Rectangle,
		wait_group: WaitGroup,
	) {
		let reference_color_option = generate_outlines(
			context.clone(),
			rectangle,
		);

		if reference_color_option.is_some() {
			fill_whole_rectangle(
				context,
				rectangle,
				&reference_color_option.unwrap(),
			);
		} else {
			split(
				context,
				rectangle,
				wait_group,
			);
		}
	}

	fn generate_outlines(
		context: GenerationContext,
		rectangle: &Rectangle,
	) -> Option<Pixel> {
		let top_y = rectangle.y;
		let bottom_y = rectangle.end_y - 1;
		let left_x = rectangle.x;
		let right_x = rectangle.end_x - 1;

		let mut outlines_identical = true;
		let reference_color = generate_point(
			context.clone(),
			left_x,
			top_y,
		);

		for x in left_x..=right_x {
			check_outline_pixel(
				context.clone(),
				x,
				top_y,
				&reference_color,
				&mut outlines_identical,
			);
			check_outline_pixel(
				context.clone(),
				x,
				bottom_y,
				&reference_color,
				&mut outlines_identical,
			);
		}
		for y in top_y..=bottom_y {
			check_outline_pixel(
				context.clone(),
				left_x,
				y,
				&reference_color,
				&mut outlines_identical,
			);
			check_outline_pixel(
				context.clone(),
				right_x,
				y,
				&reference_color,
				&mut outlines_identical,
			);
		}
		if outlines_identical {
			Option::Some(reference_color)
		} else {
			Option::None
		}
	}

	fn check_outline_pixel(
		context: GenerationContext,
		x: usize,
		y: usize,
		reference_color: &Pixel,
		outlines_identical: &mut bool,
	) {
		let current_color = generate_point(context, x, y);
		if current_color != *reference_color {
			*outlines_identical = false;
		}
	}

	fn split(
		context: GenerationContext,
		rectangle: &Rectangle,
		wait_group: WaitGroup,
	) {
		let parts = split_rectangle(rectangle);
		for part in parts.iter() {
			let child_context = context.clone();
			let child_rect = part.clone();
			let child_wait_group = wait_group.clone();
			task::spawn(async move {
				generate_within(child_context, child_rect, child_wait_group)
			});
		}
	}

	fn fill_whole_rectangle(
		context: GenerationContext,
		rectangle: &Rectangle,
		color: &Pixel,
	) {
		for x in rectangle.x..rectangle.end_x {
			for y in rectangle.y..rectangle.end_y {
				let target = context.image.index_mut_unsafe(x, y);
				*target = *color;
			}
		}
	}

	fn assert_image_matches_dimensions(image: &Image, dimensions: &ImageDimensions) {
		assert_eq!(image.height, dimensions.height);
		assert_eq!(image.width, dimensions.width);
	}

	fn rectangle_from_dimensions(dimensions: &ImageDimensions) -> Rectangle {
		Rectangle::new(
			0, 0,
			dimensions.width, dimensions.height,
		)
	}
}